rllib/algorithms/impala/tf/impala_tf_learner.py

```diff
@@ -0,0 +1,169 @@
+            "Either rollout_frag_or_episode_len or recurrent_seq_len"
+            " must be set in the IMPALA HParams. "
+        )
+
+    @override(TfLearner)
+    def compute_loss_per_module(
+        self, module_id: str, batch: SampleBatch, fwd_out: Mapping[str, TensorType]
+    ) -> TensorType:
+        values = fwd_out[SampleBatch.VF_PREDS]
+        target_policy_dist = fwd_out[SampleBatch.ACTION_DIST]
+
+        # this is probably a horribly inefficient way to do this. I should be able to
+        # compute this in a batch fashion
+        behaviour_actions_logp = batch[SampleBatch.ACTION_LOGP]
+        target_actions_logp = target_policy_dist.logp(batch[SampleBatch.ACTIONS])
+
+        behaviour_actions_logp_time_major = make_time_major(
+            behaviour_actions_logp,
+            trajectory_len=self.rollout_frag_or_episode_len,
+            recurrent_seq_len=self.recurrent_seq_len,
+            drop_last=self.vtrace_drop_last_ts,
+        )
+        target_actions_logp_time_major = make_time_major(
+            target_actions_logp,
+            trajectory_len=self.rollout_frag_or_episode_len,
+            recurrent_seq_len=self.recurrent_seq_len,
+            drop_last=self.vtrace_drop_last_ts,
+        )
+        values_time_major = make_time_major(
+            values,
+            trajectory_len=self.rollout_frag_or_episode_len,
+            recurrent_seq_len=self.recurrent_seq_len,
+            drop_last=self.vtrace_drop_last_ts,
+        )
+        bootstrap_value = values_time_major[-1]
+        rewards_time_major = make_time_major(
+            batch[SampleBatch.REWARDS],
+            trajectory_len=self.rollout_frag_or_episode_len,
+            recurrent_seq_len=self.recurrent_seq_len,
+            drop_last=self.vtrace_drop_last_ts,
+        )
+
+        # how to compute discouts?
+        # should they be pre computed?
+        discounts_time_major = (
+            1.0
+            - tf.cast(
+                make_time_major(
+                    batch[SampleBatch.TERMINATEDS],
+                    trajectory_len=self.rollout_frag_or_episode_len,
+                    recurrent_seq_len=self.recurrent_seq_len,
+                    drop_last=self.vtrace_drop_last_ts,
+                ),
+                dtype=tf.float32,
+            )
+        ) * self.discount_factor
+        vtrace_adjusted_target_values, pg_advantages = vtrace_tf2(
+            target_action_log_probs=target_actions_logp_time_major,
+            behaviour_action_log_probs=behaviour_actions_logp_time_major,
+            rewards=rewards_time_major,
+            values=values_time_major,
+            bootstrap_value=bootstrap_value,
+            clip_pg_rho_threshold=self.vtrace_clip_pg_rho_threshold,
+            clip_rho_threshold=self.vtrace_clip_rho_threshold,
+            discounts=discounts_time_major,
+        )
+
+        # The policy gradients loss.
+        pi_loss = -tf.reduce_sum(target_actions_logp_time_major * pg_advantages)
```

Comment:
Over here I'm borrowing from our original implementation. Why don't we try it out during benchmarking and see if there is a difference.

Reply:
Update: I tried it out and the learning collapsed. I couldn't find the set of optimal hparams to make it work when using mean vs sum. 

I think at this point its not a useful change, especially since we already have benchmarked this algorithm with this loss on many envs, and making this change would require rebenchmarking to find the optimal hparams.

Reply:
I mean you could extract batch_size (B) and then do only one tf.reduce_sum(xxxx) and then do loss = sum / B in order to not do the same computation twice.

Reply:
fixed